### ISSUE TITLE
[RFC] provide Python/R implementations of all the built-in objectives?

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#6440]
## Summary

Should we provide example Python (and maybe R) implementations of LightGBM's objective functions which exactly match the behavior of the builtin objectives from the C++ side?

## Motivation

Over the years of maintaining LightGBM, I've seen significant interest in implementing LightGBM's built-in objective functions in Python, for purposes like:

* learning how LightGBM works (for people who are not comfortable with C++)
* making it easier to measure the difference between custom objectives and LightGBM builtin ones
  - *(e.g. if you have a Python function that exactly matches the builtin, then you can modify it and know any performance differences are due to your modifications)*

See "References" for evidence.

## Description

I am **NOT** proposing adding such implementations to any library that we publish.

Instead, I'm thinking of something like the following:

* new directory in `examples/` containing these implementations
* tests that run in CI which compare the results to those calculated by the C++ side
* those implementations accounting for the main concerns that confuse people:
   - calculating an `init_score` if `Dataset` doesn't have one
   - correctly using sample weights
   - correctly respecting `boost_from_average`

Things that do not necessarily need to be in scope for the first versions of implementations:

* distributed training / collective operations
* respect for `deterministic` parameter
* anything related to quantized training
* exact numerical precision *(being within, say, `1e-6`, would probably good enough to start)*

## References

GitHub posts that could be summarized as "how do I replicate a built-in LightGBM objective in Python"?

* #6219
* #6160
* #6145
* #6062
* #6041
* #6040
* #5839
* #5735
* #5543
* #5524
* #5373
* #5350
* #5256
* #5114
* #5043
* #4981
* #4211
* #4077
* #3532
* #3312
* #3052
* #2834
* #2239
* #2182

And Stack Overflow:

* ["Custom L1 loss objective in LGBMRegressor"](https://stackoverflow.com/questions/77849248/custom-l1-loss-objective-in-lgbmregressor)
* ["How to reproduce LGBM (lightGBM) poisson loss in customized objective function"](https://stackoverflow.com/questions/68883433/how-to-reproduce-lgbm-lightgbm-poisson-loss-in-customized-objective-function)
* ["Custom loss function of tweedie/Regression_l1 objective in LGBMRegressor python"](https://stackoverflow.com/questions/77365771/custom-loss-function-of-tweedie-regression-l1-objective-in-lgbmregressor-python)

**Comment by shiyu1994** [comment]
I agree with this proposal. But we should also note that there can be minor (numerical) differences between Python (or R) and c++ implementations that can cause slight different results.

**Comment by jameslamb** [comment]
Linking another relevant discussion with lots of helpful information from @jmoralez : https://github.com/microsoft/LightGBM/issues/6649#issuecomment-2361796522

**Comment by Phimos** [comment]
I support this proposal. I've looked at the source code related to ObjectiveFunction. But do we need to expose the `BoostFromScore` and `RenewTreeOutput` methods to Python in order to implement this?

**Comment by jameslamb** [comment]
> But do we need to expose the BoostFromScore and RenewTreeOutput methods to Python in order to implement this?

No, this issue is not about exposing the ability to invoke the builtin objective function (implemented in C++) independently on a dataset. That's being proposed in #6586 if you want to follow along.

This is about providing pure-Python and pure-R implementations of those objective functions that can be passed into `lgb.train()` and similar, to serve as a starting point for researchers or other users wanting to customize LightGBM with domain-specific objective functions.

**Comment by Phimos** [comment]
> This is about providing pure-Python and pure-R implementations of those objective functions that can be passed into `lgb.train()` and similar, to serve as a starting point for researchers or other users wanting to customize LightGBM with domain-specific objective functions.

However, for the L1 objective function, if the `RenewTreeOutput` method cannot be overwritten, even if the same objective function is passed to `lgb.train()`, the same training results cannot be obtained.
